### PR TITLE
Add compound class to layout wrapper for global spacing styles

### PR DESCRIPTION
**Pull request by tellthemachines**
## What?

In #45326 I found that adding spacing support to Cover block didn't work out of the box, due to Cover's nested markup structure. As of #44600, layout classnames are added to the block's inner wrapper, so the `.wp-block-cover.is-layout-flow` selector used for block spacing won't work.

Because spacing styles need to target only the direct descendants of the layout wrapper, simply separating the selectors e.g. `.wp-block-cover .is-layout-flow` also won't work. 

We need a selector that targets the layout wrapper for each particular block type. This PR implements that selector, and adds spacing support to the Cover block as a testable example.

Note: I deliberately avoided any solution that would require specifying a target wrapper classname in the layout object, because once #47807 is done it should be possible to enable/disable layout on any block from theme.json, without knowledge of the block's markup structure. Also, the whole point of  #44600 was to avoid doing precisely that 😄 

## Testing Instructions

1. In global styles, set a block spacing value for Cover block and another for Group block;
2. Add a Cover to a post or template, add a couple of Paragraphs and a Group block to it, then a couple more Paragraphs inside the Group block;
3. Check that the correct spacing values are used on both Cover and Group block children;
4. Set a different block spacing value in the Cover block settings and check that it overrides the global value.

## Screenshots or screencast 

Front end with global spacing:

Screenshot 2023-02-10 at 3 01 08 pm

Block editor settings:

Screenshot 2023-02-10 at 2 59 39 pm

### CODE REVIEW

File: lib/block-supports/layout.php
Comment (andrewserong) [inline lib/block-supports/layout.php:474]: Can this logic be replaced by a call to `wp_get_block_default_classname` so that any applied filters to the classname can be applied before we concatenate it with the layout classname suffix? E.g. something like:

```
$class_names[] = wp_get_block_default_classname( $block['blockName'] ) . '-' . $layout_classname;
```

File: packages/block-editor/src/hooks/layout.js
Comment (andrewserong) [inline packages/block-editor/src/hooks/layout.js]: Similar to the comment on the PHP code, can we use `getBlockDefaultClassName( name )` here to get the classname?

File: lib/class-wp-theme-json-gutenberg.php
Comment (andrewserong) [inline lib/class-wp-theme-json-gutenberg.php]: Sorry if this is obvious, but I wasn't too sure: why does this need the `:root` prefix, is it to increase specificity now that it's a dash, instead of a combined classname selector? And is `:root` preferable over using `static::ROOT_BLOCK_SELECTOR`?

**Review by andrewserong — COMMENTED** [review]
Very clever idea @tellthemachines, nice solution for working around the idea of blocks having to explicitly declare the class for the inner wrapper. I think concatenating the block classname with the layout type is a good way to go about it, and makes the output from global styles quite explicit, too. To me, this feels like a good part of the overall stabilisation of the layout support, in addition to fixing the issue at hand 👍

From a quick test, it appears to be working well for me so far, with setting block-level spacing for Cover and Social icons blocks in global styles, with real-block-level attributes overriding the global styles values, etc.

Just left a couple of comments about seeing if we can do a lookup for the block classname instead, and a question about the addition of `:root` as a prefix in the rules output.

Overall, though, I think this sounds like a good direction to me!

File: lib/block-supports/layout.php
Comment (tellthemachines) [reply lib/block-supports/layout.php:474]: Good question! That could result in a mismatch between classnames in markup and CSS output, because the CSS rules are using [a similar logic](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L911) to generate the block classname. 

Might the existence of classname filters have any impact on this functionality? I can't find any use of this filter in Core, do you know of any examples?

File: lib/class-wp-theme-json-gutenberg.php
Comment (tellthemachines) [reply lib/class-wp-theme-json-gutenberg.php]: I'm pretty sure I added it for specificity, but am no longer sure of the actual rule it was meant to override 😅 I can try removing it and see how we go. Good idea to use `static::ROOT_BLOCK_SELECTOR` if we do end up needing it!

File: lib/block-supports/layout.php
Comment (andrewserong) [reply lib/block-supports/layout.php:474]: > do you know of any examples?

No, I don't know of any examples where folks are using the filter. It could be fairly uncommon?

> That could result in a mismatch between classnames in markup and CSS output, because the CSS rules are using [a similar logic](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L911) to generate the block classname.

If we switched to using `wp_get_block_default_classname`, I imagine we'd also switch to using `getBlockDefaultClassName( name )` in the JS logic, so there (at least theoretically) shouldn't be a mismatch. Both functions share the same logic, and it's pretty similar to the logic in this PR, I think, from a quick skim of the function definitions.

The main benefit of using those functions probably isn't really the filtering, though, it's probably just a small (and non-blocking) opportunity to re-use existing functions if it feels like it'd neaten things. But you do raise a good point about ensuring consistency between the JS and PHP output! I don't feel too strongly either way, as it'd be a quick change if we want to pick one approach and later flip to the other if we change our minds 🙂

File: lib/class-wp-theme-json-gutenberg.php
Comment (tellthemachines) [reply lib/class-wp-theme-json-gutenberg.php]: Ok, so a bit of testing shows that we actually need `:root` to override the default layout styles, because it has a slightly higher specificity than `body`:

Screenshot 2023-02-14 at 4 55 36 pm

Leaving it as is for now.

File: lib/block-supports/layout.php
Comment (tellthemachines) [reply lib/block-supports/layout.php:474]: It would be fine to use in the layout JS logic, it's in the global styles output that things get dodgy because where we're generating the CSS rule around [here](https://github.com/WordPress/gutenberg/blob/add/cover-spacing/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L398), the selector might not be a block at all. And if it is a block, its selector was generated using the [same logic](https://github.com/WordPress/gutenberg/blob/add/cover-spacing/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L911) I used here 😅 

I'm not sure if that bit can be replaced by `wp_get_block_default_classname` without any side effects, might have to create a classname filter to test it.

Comment (andrewserong) [reply lib/block-supports/layout.php:474]: Ah gotcha! In that case, maybe it's better to stick with the rules you've got in this PR for now, if it looks like it's going to be either brittle or tricky to get it to work in the global styles logic 👍

File: lib/class-wp-theme-json-gutenberg.php
Comment (andrewserong) [reply lib/class-wp-theme-json-gutenberg.php]: Thanks for confirming!

I had another idle thought — what if we added the layout classname to the selector rather than the `:root ` prefix, so that the rule becomes `.wp-block-group-is-layout-contrained.is-layout-constrained` — I think this would then have the same level of specificity as before this PR. On the other hand, that's also an unwieldy looking selector 😅

I'm quite possibly overthinking it, but I quite like the idea of the block-level-in-global-styles rules being a kind of combination of the block's classname and the layout classname, if we can get it to work, without having to add `:root ` in order to increase the specificity. My main concern about the `:root ` prefix is that while it works, it wasn't obvious to me from looking at the rule that it's there for the extra specificity, whereas combining classnames seems a little clearer... but that could just be me (and definitely my naivety about the `:root` pseudo-class 🙂), so I don't feel too strongly about it.

Comment (tellthemachines) [reply lib/class-wp-theme-json-gutenberg.php]: Hmm that's a good point. I _think_ `:root` has the same specificity as a regular classname, so we could do something like that, or even duplicate the composite selector like `.wp-block-group-is-layout-contrained.wp-block-group-is-layout-contrained`, so it's even more obvious that the bump is for specificity.

Comment (tellthemachines) [reply lib/class-wp-theme-json-gutenberg.php]: I'm not sure how this is going to pan out relative to #47858 though 😅

Comment (andrewserong) [reply lib/class-wp-theme-json-gutenberg.php]: > I'm not sure how this is going to pan out relative to https://github.com/WordPress/gutenberg/pull/47858 though 😅

Good point! I reckon we'll likely be doing a fair bit of nudging back and forth before we land on a sweet spot 🤞 😄

**Comment by apeatling** [comment]
**Edit: Never mind on this! I remembered we don't actually do this yet. It's tripped me up before because you'd 100% expect to see the setting show up as pre-configured.**

Started testing this, but humour me for a second. If I set a non-custom block spacing on the cover block in global styles, it does not seem to apply to a cover block inserted into a post? Is that working for you? I've tried multiple saves etc. The global styles setting sticks for cover, but does not apply to a cover block in a new post.

Set in global styles specifically for the cover block:
Screenshot 2023-03-06 at 10 51 53 AM

New cover block in a post:
Screenshot 2023-03-06 at 10 52 04 AM

**Comment by tellthemachines** [comment]
> If I set a non-custom block spacing on the cover block in global styles, it does not seem to apply to a cover block inserted into a post?

The global styles value should apply to the block, because the generated styles are loaded in the editor, but it won't be visible in the block's spacing control because we've only recently had the ability to access the actual values from global styles in the post editor (since #47098) and we haven't updated the design tools to show them yet 😅

**Comment by apeatling** [comment]
> The global styles value should apply to the block, because the generated styles are loaded in the editor, but it won't be visible in the block's spacing control because we've only recently had the ability to access the actual values from global styles in the post editor (since https://github.com/WordPress/gutenberg/pull/47098) and we haven't updated the design tools to show them yet 😅

Thanks for the extra context, everything worked fine. It was simply the control not showing the global value that threw me off. I think we should work on that now that the ability is there since it's something that connects dots for people, and seems broken if you don't know the context.

File: lib/block-supports/layout.php
Comment (andrewserong) [reply lib/block-supports/layout.php:474]: Just resolving this old thread — using the hard-coded logic in this PR sounds fine to me, and ensures there's consistency without filters messing around with things unexpectedly 😄

**Review by andrewserong — APPROVED** [review]
Nice work @tellthemachines, this looks like it's in a good place to me, especially since this PR is targeting being merged into the layout support for the Cover block PR! If (eventually) there's ever any concern about the additional classnames being injected in blocks that don't have global styles rules applied, we could always look into a potential follow up that only outputs the additional classname if a rule exists, but I suspect that won't be necessary.

✅ Re-tested that block level gap in global styles is output correctly in the site editor, post editor, and site front-end
✅ Block-level spacing rules still override those set at the block-level in global styles
✅ Fallback / default spacing rules in classic themes still work as-is (e.g. default flex gap that gets applied to social icons blocks)

LGTM! ✨

**Comment by tellthemachines** [comment]
Thanks for reviewing @andrewserong! I won't merge this into the Cover block PR until that one's approved, given that it will make it harder to review. We can merge it directly into trunk if we need to.

**Comment by tellthemachines** [comment]
## Dev note

### Compound block + layout type classname applied to inner wrapper of layout blocks

As of WP 6.2, [layout classnames are added to the block inner wrapper](https://github.com/WordPress/gutenberg/pull/44600). For most blocks, this is the same as the outer wrapper, but some blocks (e.g. core Cover) have multiple inner containers. In 6.3, a new classname is added to the inner wrapper of all blocks with layout, comprised of block classname + layout classname, e.g.: `.wp-block-cover-is-layout-constrained`.
This makes it possible for  blocks with a complex markup structure to support custom spacing styles.

**Pull request merged**
